src/main/java/com/walmartlabs/x12/common/AbstractStandardX12Parser.java

```diff
@@ -0,0 +1,118 @@
+/**
+Copyright (c) 2018-present, Walmart, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+ */
+package com.walmartlabs.x12.common;
+
+import com.walmartlabs.x12.X12Parser;
+import com.walmartlabs.x12.X12Segment;
+import com.walmartlabs.x12.exceptions.X12ParserException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.StringUtils;
+
+import java.util.List;
+
+public abstract class AbstractStandardX12Parser<T extends AbstractStandardX12Document> implements X12Parser<T> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractStandardX12Parser.class);
+
+    public static final String ISA_ID = "ISA";
+    public static final String GROUP_SEGMENT_ID = "GS";
+
+    /**
+     * template for parsing a standard EDI X12 document
+     *
+     * @return T (the Class associated with the parser)
+     * @throws X12ParserException
+     */
+    @Override
+    public T parse(String sourceData) {
+        AbstractStandardX12Document x12Doc = null;
+
+        if (!StringUtils.isEmpty(sourceData)) {
+            x12Doc = this.createX12Document();
+            List<X12Segment> segmentLines = this.splitSourceDataIntoSegments(sourceData);
+            int segmentIdx = 0;
+            this.parseInterchangeControlHeader(segmentLines.get(segmentIdx++), x12Doc);
+            this.parseGroupHeader(segmentLines.get(segmentIdx++), x12Doc);
+            this.parseCustom(segmentLines, x12Doc);
+        }
+
+        return (T) x12Doc;
+    }
+
+    protected abstract AbstractStandardX12Document createX12Document();
+
+    protected abstract void parseCustom(List<X12Segment> segmentLines, AbstractStandardX12Document x12Doc);
+
+    /**
+     * parse the ISA segment
+     * @param segment
+     * @param x12
+     */
+    protected void parseInterchangeControlHeader(X12Segment segment, AbstractStandardX12Document x12Doc) {
+        LOGGER.debug(segment.getSegmentIdentifier());
+
+        String segmentIdentifier = segment.getSegmentIdentifier();
+        if (ISA_ID.equals(segmentIdentifier)) {
+            InterchangeControlHeader isa = new InterchangeControlHeader();
+            isa.setAuthorizationInformationQualifier(segment.getSegmentElement(1));
```

Comment:
Since I don't have the manual, I am trusting you here(setting the values correctly to `InterchangeControlHeader` and `GroupHeader`) 😛

Reply:
I think it is still uploading to your account 
![image](https://user-images.githubusercontent.com/12648091/54031110-c7c6fb00-417b-11e9-8f0b-8d60ad2729b9.png)